File: services/backend/BlazorApp.Shared/Models/HBweb/LocalSupplierCategory.cs
using BlazorApp.Shared.Helper;
using SqlSugar;

namespace BlazorApp.Shared.Models
{
    /// <summary>
    /// 本地供应商自己的商品分类（来自供应商网站，由浏览器扩展采集）。
    /// 供应商 200（Hot Bargain 自营）不使用本表，其供应商分类即仓库分类 <see cref="WarehouseCategory"/>。
    /// </summary>
    [SugarTable("LocalSupplierCategory")]
    public class LocalSupplierCategory : BaseEntity
    {
        [SugarColumn(IsPrimaryKey = true, IsNullable = false, Length = 50)]
        public string CategoryGUID { get; set; } = UuidHelper.GenerateUuid7();

        [SugarColumn(IsNullable = false, Length = 64)]
        public string LocalSupplierCode { get; set; } = string.Empty;

        [SugarColumn(IsNullable = true, Length = 50)]
        public string? ParentGUID { get; set; }

        [SugarColumn(IsNullable = false, Length = 200)]
        public string CategoryName { get; set; } = string.Empty;

        /// <summary>
        /// 站点稳定标识：归一化后的分类页路径（可附带白名单查询参数），同一供应商内唯一。
        /// </summary>
        [SugarColumn(IsNullable = false, Length = 400)]
        public string ExternalKey { get; set; } = string.Empty;

        /// <summary>
        /// 反规范化的完整路径 "A &gt; B &gt; C"，列表展示无需回溯祖先。
        /// </summary>
        [SugarColumn(IsNullable = false, Length = 1000)]
        public string FullPath { get; set; } = string.Empty;

        /// <summary>
        /// 层级深度，根为 0；自动归类取最深的非促销分类。
        /// </summary>
        [SugarColumn(IsNullable = false)]
        public int Depth { get; set; }

        [SugarColumn(IsNullable = true, Length = 1000)]
        public string? SourceUrl { get; set; }

        /// <summary>
        /// 促销/横切分类（Clearance、New 等），不参与自动归类。
        /// </summary>
        [SugarColumn(IsNullable = false)]
        public bool IsPromotional { get; set; }

        /// <summary>
        /// 促销标记来源：pattern（配置规则判定）或 manual（人工切换，重算时不再覆盖）。
        /// </summary>
        [SugarColumn(IsNullable = false, Length = 16)]
        public string PromotionalSource { get; set; } = LocalSupplierCategoryPromotionalSources.Pattern;

        [SugarColumn(IsNullable = true)]
        public int? SortOrder { get; set; }

        [SugarColumn(IsNullable = false)]
        public bool IsActive { get; set; } = true;

        [SugarColumn(IsNullable = false)]
        public DateTime FirstSeenAt { get; set; } = DateTime.UtcNow;

        [SugarColumn(IsNullable = false)]
        public DateTime LastSeenAt { get; set; } = DateTime.UtcNow;
    }

    public static class LocalSupplierCategoryPromotionalSources
    {
        public const string Pattern = "pattern";
        public const string Manual = "manual";
    }
}

File: apps/web/src/types/localSupplierCategory.ts
/**
 * 供应商分类（分类数据来自供应商网站）的前端契约。
 * 后端 System.Text.Json 全局 camelCase；服务层把响应归一化成这里的形状。
 */

/** Hot Bargain 自营供应商：它的供应商分类就是仓库分类，不走网站采集。 */
export const HOT_BARGAIN_SUPPLIER_CODE = '200'

/** 商品上的供应商分类来源：网站采集自动归类 / 人工指定并锁定 / 200 随仓库分类。 */
export type SupplierCategorySource = 'website' | 'manual' | 'warehouse'

export const SUPPLIER_CATEGORY_SOURCES: readonly SupplierCategorySource[] = ['website', 'manual', 'warehouse']

export type LocalSupplierCategorySourceKind = 'warehouse' | 'website'

export type LocalSupplierCategoryPromotionalSource = 'pattern' | 'manual'

/** 管理弹窗左侧的每供应商统计（GET summary）。 */
export interface LocalSupplierCategorySummary {
  supplierCode: string
  supplierName?: string
  sourceKind: LocalSupplierCategorySourceKind
  categoryCount: number
  promotionalCount: number
  productCount: number
  assignedCount: number
  manualCount: number
  unassignedCount: number
  lastCapturedAt?: string
  lastSnapshotAt?: string
}

/** 供应商分类树节点（GET tree?supplierCode=）。200 时由服务端把仓库分类树映射成同一结构。 */
export interface LocalSupplierCategoryNode {
  categoryGuid: string
  parentGuid?: string
  name: string
  externalKey?: string
  fullPath?: string
  depth: number
  isPromotional: boolean
  promotionalSource?: LocalSupplierCategoryPromotionalSource
  isActive: boolean
  sortOrder?: number
  sourceUrl?: string
  productCount: number
  lastSeenAt?: string
  children: LocalSupplierCategoryNode[]
}

/** PATCH {categoryGuid}/promotional 的结果。 */
export interface LocalSupplierCategoryPromotionalResult {
  reassigned: number
  cleared: number
}

/** POST {supplierCode}/resolve 的结果。 */
export interface LocalSupplierCategoryResolveResult {
  productsScanned: number
  assigned: number
  updated: number
  cleared: number
  unchanged: number
  manualSkipped: number
  staleRemoved: number
}

/**
 * 商品单个/批量更新时的供应商分类三态：
 * - 带 supplierCategoryGUID：人工指定并锁定；
 * - clearSupplierCategory=true：删除人工/自动归属并按采集数据重新自动归类；
 * - 两者都不带：不变。
 */
export interface SupplierCategoryUpdatePayload {
  supplierCategoryGUID?: string
  clearSupplierCategory?: boolean
}

/** 人工指定分类与所选商品供应商不一致时后端返回的错误码。 */
export const SUPPLIER_CATEGORY_MISMATCH_ERROR_CODE = 'CATEGORY_SUPPLIER_MISMATCH'
